src/components/MapView/DateSelector/index.tsx

```diff
@@ -43,38 +51,148 @@ function DateSelector({ availableDates = [], classes }: DateSelectorProps) {
+      value: 0,
+      label: '',
+      month: '',
+      isFirstday: false,
+    },
+  ]);
+
+  const [timelinePosition, setTimelinePosition] = useState({ x: 0, y: 0 });
+  const [pointerPosition, setPointerPosition] = useState({ x: 0, y: 0 });
+
+  const dateRef = useRef(availableDates);
 
-  const selectedYear = selectedDate.year();
+  const timeLine = useRef(null);
+  const timeLineWidth = get(timeLine.current, 'offsetWidth', 0);
 
+  // Move the slider automatically so that the pointer always visible
   useEffect(() => {
+    let x = 0;
+    if (pointerPosition.x >= dateRange.length * 10 - timeLineWidth) {
+      // eslint-disable-next-line fp/no-mutation
+      x = timeLineWidth - dateRange.length * 10;
+    } else if (pointerPosition.x > timeLineWidth) {
+      // eslint-disable-next-line fp/no-mutation
+      x = -1 * pointerPosition.x + timeLineWidth / 2;
+    }
+    if (
+      -1 * timelinePosition.x > pointerPosition.x ||
+      -1 * timelinePosition.x + timeLineWidth < pointerPosition.x
+    ) {
+      setTimelinePosition({ x, y: 0 });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [pointerPosition]);
+
+  // Create timeline range and set pointer position
+  useEffect(() => {
+    const range = Array.from(
+      moment()
+        .range(
+          moment(stateStartDate).startOf('year'),
+          moment(stateStartDate).endOf('year'),
+        )
+        .by('days'),
+    ).map(e => {
+      return {
+        value: e.valueOf(),
+        label: e.format('DD MMM YYYY'),
+        month: e.format('MMM YYYY'),
+        isFirstday: e.date() === e.startOf('month').date(),
+      };
+    });
+    setDateRange(range);
+    const dateIndex = findIndex(range, e => {
+      return e.label === moment(stateStartDate).format('DD MMM YYYY');
+    });
+    setPointerPosition({
+      x: dateIndex * 10,
+      y: 0,
+    });
     setSelectedDate(moment(stateStartDate));
   }, [stateStartDate]);
-  function updateMonth(month: number) {
-    const { startDate, endDate } = getMonthStartAndEnd(month, selectedYear);
-    dispatch(updateDateRange({ startDate, endDate }));
-  }
-  function incrementYear() {
-    setSelectedDate(selectedDate.clone().year(selectedYear + 1));
-  }
-  function decrementYear() {
-    setSelectedDate(selectedDate.clone().year(selectedYear - 1));
-  }
+
   function updateStartDate(date: Date) {
     const time = date.getTime();
     dispatch(updateDateRange({ startDate: time, endDate: time }));
   }
 
+  function setDatePosition(date: number, increment: number) {
+    const dates = availableDates.map(d => {
+      return d + USER_DATE_OFFSET;
+    });
+    const selectedIndex = findDateIndex(dates, date as number);
+    if (dates[selectedIndex + increment]) {
+      updateStartDate(new Date(dates[selectedIndex + increment]));
+    }
+  }
+
+  // move pointer to closest date when change map layer
+  if (!isEqual(dateRef.current, availableDates)) {
+    setDatePosition(stateStartDate as number, 0);
```

Comment:
Its a little unusual to have `setState` outside of a `useEffect` or function. We should also not assume stateStartDate isn't undefined, since the typing tells us it could be.

Reply:
Maybe making the function setDatePosition handle undefined dates correctly would remove the need to cast.